protobuf/lib/src/protobuf/coded_buffer.dart

```diff
@@ -199,52 +280,45 @@ void _mergeFromCodedBufferReader(BuilderInfo meta, _FieldSet fs,
   }
 }
 
-void _readPackable(BuilderInfo meta, _FieldSet fs, CodedBufferReader input,
-    int wireType, FieldInfo fi, Function() readFunc) {
-  void readToList(List list) => list.add(readFunc());
-  _readPackableToList(meta, fs, input, wireType, fi, readToList);
+void _readPacked(CodedBufferReader input, void Function() readFunc) {
+  input._withLimit(input.readInt32(), () {
+    while (!input.isAtEnd()) {
+      readFunc();
+    }
+  });
 }
 
 void _readPackableToListEnum(
+    List list,
     BuilderInfo meta,
     _FieldSet fs,
     CodedBufferReader input,
     int wireType,
-    FieldInfo fi,
     int tagNumber,
     ExtensionRegistry registry) {
-  void readToList(List list) {
-    final rawValue = input.readEnum();
-    final value = meta._decodeEnum(tagNumber, registry, rawValue);
-    if (value == null) {
-      final unknown = fs._ensureUnknownFields();
-      unknown.mergeVarintField(tagNumber, Int64(rawValue));
-    } else {
-      list.add(value);
-    }
-  }
-
-  _readPackableToList(meta, fs, input, wireType, fi, readToList);
-}
-
-void _readPackableToList(
-    BuilderInfo meta,
-    _FieldSet fs,
-    CodedBufferReader input,
-    int wireType,
-    FieldInfo fi,
-    Function(List) readToList) {
-  final list = fs._ensureRepeatedField(meta, fi);
-
   if (wireType == WIRETYPE_LENGTH_DELIMITED) {
     // Packed.
     input._withLimit(input.readInt32(), () {
       while (!input.isAtEnd()) {
-        readToList(list);
+        final rawValue = input.readEnum();
+        final value = meta._decodeEnum(tagNumber, registry, rawValue);
+        if (value == null) {
+          final unknown = fs._ensureUnknownFields();
+          unknown.mergeVarintField(tagNumber, Int64(rawValue));
+        } else {
+          list.add(value);
+        }
       }
     });
   } else {
     // Not packed.
-    readToList(list);
+    final rawValue = input.readEnum();
+    final value = meta._decodeEnum(tagNumber, registry, rawValue);
+    if (value == null) {
+      final unknown = fs._ensureUnknownFields();
+      unknown.mergeVarintField(tagNumber, Int64(rawValue));
+    } else {
+      list.add(value);
+    }
```

Comment:
Why do you handle enums differently than all the others? E.g. why not do this:

```dart
        final list = cachedList ??= fs._ensureRepeatedField(meta, fi);
        if (wireType == WIRETYPE_LENGTH_DELIMITED) {
          _readPacked(input, () => readEnum(input));
        } else {
          list.add(readEnum(input));
        }
        break;
```

or handle rest of the types similarly, with functions like `_readPackableToListInt64`, `_readPackableToListSing32` etc.?

Currently enums are handled differently (more efficiently) than all the others, but I don't know why this should be the case.

Reply:
I've lost a lot of context but looking at the code I'd say it's probably because enums and ints are handled differently in the skew case. Specifically look at the "unknown fields" path:

```dart
final rawValue = input.readEnum();
final value = meta._decodeEnum(tagNumber, registry, rawValue);
if (value == null) {
  final unknown = fs._ensureUnknownFields();
  unknown.mergeVarintField(tagNumber, Int64(rawValue));
} else {
  list.add(value);
}
```

If you look at the non repeated flavors, enums are also handled differently:

```dart
case PbFieldType._OPTIONAL_DOUBLE:
  fs._setFieldUnchecked(meta, fi, input.readDouble());
  break;
case PbFieldType._OPTIONAL_ENUM:
  final rawValue = input.readEnum();
  final value = meta._decodeEnum(tagNumber, registry, rawValue);
  if (value == null) {
    final unknown = fs._ensureUnknownFields();
    unknown.mergeVarintField(tagNumber, Int64(rawValue));
  } else {
    fs._setFieldUnchecked(meta, fi, value);
  }
  break;
````

And lastly, your proposal simply breaks the benchmark